### PR TITLE
Improvements to GH Actions in our CI

### PR DESCRIPTION
**Pull request by causten**
Build and Publish publicly accessible images.  This will allow PRs from forks to work.
Improve docker building speed by using layers already built (see `--cache-from`)

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -75,11 +76,17 @@ jobs:
 
     - name: Build and publish 
       env:            
-        DOCKERIMAGE: ${{ needs.check_image.outputs.imagetag }}
+        DOCKER_TAG_UBUNTU: ${{ needs.check_image.outputs.imagetag }}
       run: |
+        # The TOKEN and USERID are github secrets, Action failures at this step
+        # can come from a PR from a fork changing a file wichi forces a rebuild
```
Comment (umangyadav) [inline .github/workflows/ci.yaml:82]:
```suggestion
        # can come from a PR from a fork changing a file which forces a rebuild
```

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -75,11 +76,17 @@ jobs:
...
+        # Resolve by making an internal PR of the Forked PR
         echo $DOCKER_TOKEN | docker login -u $DOCKER_USER --password-stdin
-        docker build . --file hip-clang.docker --tag $DOCKERIMAGE;
-        docker push $DOCKERIMAGE;
+
+        docker pull $DOCKER_IMAGE_UBUNTU:latest || true
+        docker build . --file hip-clang.docker --cache-from $DOCKER_IMAGE_UBUNTU:latest --tag $DOCKER_IMAGE_UBUNTU:$DOCKER_TAG_UBUNTU --tag $DOCKER_IMAGE_UBUNTU:latest;
```
Comment (umangyadav) [inline .github/workflows/ci.yaml:87]: Why two tags ?

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -75,11 +76,17 @@ jobs:
...
+        DOCKER_TAG_UBUNTU: ${{ needs.check_image.outputs.imagetag }}
       run: |
+        # The TOKEN and USERID are github secrets, Action failures at this step
+        # can come from a PR from a fork changing a file wichi forces a rebuild
+        # Resolve by making an internal PR of the Forked PR
         echo $DOCKER_TOKEN | docker login -u $DOCKER_USER --password-stdin
-        docker build . --file hip-clang.docker --tag $DOCKERIMAGE;
-        docker push $DOCKERIMAGE;
+
+        docker pull $DOCKER_IMAGE_UBUNTU:latest || true
```
Comment (umangyadav) [inline .github/workflows/ci.yaml:86]: what `|| true` does here ? i haven't seen such thing before.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -90,18 +97,24 @@ jobs:
     - uses: actions/checkout@v3
     - name: Build and publish SLES
       env:            
-        DOCKERIMAGE_SLES: ${{ needs.check_image.outputs.imagetag_sles }}
+        DOCKER_TAG_SLES: ${{ needs.check_image.outputs.imagetag_sles }}
       run: |
+        # The TOKEN and USERID are github secrets, Action failures at this step
+        # can come from a PR from a fork changing a file wichi forces a rebuild
```
Comment (umangyadav) [inline .github/workflows/ci.yaml:103]:
```suggestion
        # can come from a PR from a fork changing a file which forces a rebuild
```

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (causten) [reply .github/workflows/ci.yaml:86]: That forces the return code to be success, there may not be a :latest tag and that would have caused an error.  It's ok if the :latest doesn't exist so just continue

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (causten) [reply .github/workflows/ci.yaml:87]: It builds the image with a hash and with a known tag named "latest".  That is so later builds can use the --cache-from xxx:latest to use existing layers.  I wouldn't know the name of an older calculated hash, so just creating a well known tag label

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (causten) [reply .github/workflows/ci.yaml:103]: Fixed

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (umangyadav) [reply .github/workflows/ci.yaml:86]: Looks like it pushes both tags 
docker push $DOCKER_IMAGE_UBUNTU:$DOCKER_TAG_UBUNTU;
docker push $DOCKER_IMAGE_UBUNTU:latest;

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (umangyadav) [reply .github/workflows/ci.yaml:87]: I think i get it.  It is helpful for the case when hash changes.

**Review by umangyadav — APPROVED** [review]

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -75,11 +76,17 @@ jobs:
...
+        DOCKER_TAG_UBUNTU: ${{ needs.check_image.outputs.imagetag }}
       run: |
+        # The TOKEN and USERID are github secrets, Action failures at this step
+        # can come from a PR from a fork changing a file which forces a rebuild
+        # Resolve by making an internal PR of the Forked PR
         echo $DOCKER_TOKEN | docker login -u $DOCKER_USER --password-stdin
-        docker build . --file hip-clang.docker --tag $DOCKERIMAGE;
-        docker push $DOCKERIMAGE;
+
+        docker pull $DOCKER_IMAGE_UBUNTU:latest || true
```
Comment (pfultz2) [inline .github/workflows/ci.yaml:86]: Is it necessary to pull this image?

**Review by TedThemistokleous — APPROVED** [review]

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (causten) [reply .github/workflows/ci.yaml:86]: From the documentation I've read it needs to be a local cache unless we enable BuildKit.

**Pull request merged**